File: Stacks_Queues/go/DoublyLinkedList.go
package types

import (
  "fmt"
)


type DoublyLinkedListNode struct {
  Data int
  Next *DoublyLinkedListNode
  Prev *DoublyLinkedListNode
}


type DoublyLinkedList struct {
  Head *DoublyLinkedListNode
  Tail *DoublyLinkedListNode
}


func (list *DoublyLinkedList) Append(data int) {
  newNode := DoublyLinkedListNode{Data: data}
  if list.Head == nil {
    list.Head = &newNode
    list.Tail = &newNode
  } else {
    prev := list.Tail
    list.Tail.Next = &newNode
    list.Tail = list.Tail.Next
    list.Tail.Prev = prev
  }
}


func (list *DoublyLinkedList) AddToBeginning(data int) {
  newNode := DoublyLinkedListNode{Data: data}
  if list.Head == nil {
    list.Head = &newNode
    list.Tail = &newNode
  } else {
    head := list.Head
    list.Head = &newNode
    head.Prev = &newNode
    newNode.Next = head
  }
}


func (list *DoublyLinkedList) DeleteLast() {
  last := list.Tail
  last.Prev.Next = nil
  list.Tail = last.Prev
}


func PrintDoublyLinkedList(list DoublyLinkedList) {
  node := list.Head
  for node != nil {
    fmt.Print(node.Data, "->")
    node = node.Next
  }
  fmt.Print("nil\n")
}

File: Stacks_Queues/python/DoublyLinkedList.py
class Node(object):

    def __init__(self, data=None, next=None, prev=None):
        self.data = data
        self.next = next
        self.prev = prev

    def __str__(self):
        return str(self.data)


class DoublyLinkedList(object):
    def __init__(self):
        self.head = None
        self.tail = None

    def __str__(self):
        return '->'.join([str(data) for data in self])

    def __iter__(self):
        current = self.head
        while current:
            yield current
            current = current.next

    def __len__(self):
        length = 0
        for value in self:
            length += 1

        return length

    def append(self, data):
        if self.head is None:
            self.head = self.tail = Node(data)
        else:
            prev = self.tail
            self.tail.next = Node(data)
            self.tail = self.tail.next
            self.tail.prev = prev

    def delete_last(self):
        last = self.tail
        last.prev.next = None
        self.tail = last.prev
